### PR TITLE
In-memory certificates data improvement

### PR DESCRIPTION
**Pull request by pv3nturi**
Improved certificates data management on memory:

- serial number, expiring date and keystore data are now keep in memory and not more retrived from decoded chain

### CODE REVIEW

File: carapace-server/src/main/java/org/carapaceproxy/server/certificates/DynamicCertificatesManager.java [hunk 1]
```diff
@@ -273,11 +284,10 @@ private void certificatesLifecycle() {
                 .sorted((e1, e2) -> e1.getKey().compareTo(e2.getKey()))
                 .map(e -> e.getValue())
                 .collect(Collectors.toList());
-        for (CertificateData data : _certificates) {
+        for (CertificateData cert : _certificates) {
             boolean updateCertificate = true;
-            final String domain = data.getDomain();
+            final String domain = cert.getDomain();
             try {
-                CertificateData cert = loadOrCreateDynamicCertificateForDomain(domain, data.isWildcard(), false, data.getDaysBeforeRenewal());
```
Comment (pv3nturi) [inline carapace-server/src/main/java/org/carapaceproxy/server/certificates/DynamicCertificatesManager.java:280]: We skip retriving the certificate from db which is already up-to-date. (all certs are reloaded as soon as a node triggers fire-cache request)

File: carapace-server/src/main/java/org/carapaceproxy/utils/CertificatesUtils.java [hunk 2]
```diff
@@ -146,19 +146,14 @@ public static KeyStore loadKeyStoreData(byte[] data, String password)
...
-        if (chain == null || chain.length == 0) {
+    public static boolean isCertificateExpired(Date expiringDate, int daysBeforeRenewal) throws GeneralSecurityException {
+        if (expiringDate == null) {
             return false;
         }
-        try {
-            Calendar cal = Calendar.getInstance();
-            cal.setTime(new Date());
-            cal.add(Calendar.DATE, daysBeforeRenewal);
-            ((X509Certificate) chain[0]).checkValidity(cal.getTime());
```
Comment (pv3nturi) [inline carapace-server/src/main/java/org/carapaceproxy/utils/CertificatesUtils.java:157]: This checks notBefore and notAfter. Maybe we just need notAfter.

**Review by nicoloboschi — CHANGES_REQUESTED** [review]
It looks good, I left a comment on the expire date check

**Review by nicoloboschi — APPROVED** [review]
+1

**Pull request merged**